submariner-operator/issue-79-kubeconfig-handling.md

```diff
@@ -0,0 +1,120 @@
+## Epic Description
+
+The various `subctl` commands support a mixture of `KUBECONFIG` environment variables, `--kubeconfig` flags, `--kubecontext` flags,
+kubeconfig files etc.
+
+We should make this consistent, ideally using `clientcmd`’s tools to handle kubeconfig.
+[Operator PR #23](https://github.com/skitt/submariner-operator/pull/23) has an old attempt at this,
+using [`BindOverrideFlags()`](https://pkg.go.dev/k8s.io/client-go@v0.23.5/tools/clientcmd#BindOverrideFlags) to provide access
+to all the features supported by `clientcmd` (see `kubectl options`).
+
+On top of that, we should align our flags with `kubectl`’s;
+beyond helping users already familiar with `kubectl` flags,
+this will make it easier to turn `subctl` into a `kubectl` plugin,
+should we choose to do so.
+This would involve turning `--kubecontext` into `--context`;
+`--kubecontext` would be preserved as a deprecated alias for `--context` for one release,
+and removed in the following release.
+
+## Acceptance Criteria
+
+For commands using a single context, _all_ the following should work,
+with kubeconfigs containing one or more clusters and/or contexts:
+
+```bash
+# Use the default context in the kubeconfig
+KUBECONFIG=/path/to/kubeconfig subctl foo
+subctl foo --kubeconfig /path/to/kubeconfig
+subctl --kubeconfig /path/to/kubeconfig foo
+
+# Use the specified context from the kubeconfig
+KUBECONFIG=/path/to/kubeconfig subctl foo --context bar
+subctl foo --kubeconfig /path/to/kubeconfig --context bar
+subctl --kubeconfig /path/to/kubeconfig --context bar foo
+```
+
+For commands using multiple contexts,
+there should be one “main” set of flags, with no prefix,
+and another set with a prefix describing the purpose of the context:
+
+```bash
+KUBECONFIG=/path/to/kubeconfig subctl benchmark latency --context foo --tocontext bar
+subctl benchmark latency --kubeconfig /path/to/kubeconfig --context foo --tocontext bar
+```
+
+`--kubeconfig` is preserved as-is, since users can use a single file combining all their contexts.
+However such commands should also support separate kubeconfigs, to allow using kubeconfigs which
+contain conflicting configuration (_e.g._ kubeconfigs using the same context name and user name,
+as produced by the OpenShift installer):
+
+```bash
+subctl benchmark latency --kubeconfig /path/to/fromconfig --toconfig /path/to/toconfig
```

Comment:
```suggestion
subctl benchmark latency --kubeconfig /path/to/fromconfig --toconfig /path/to/toconfig --context foo --tocontext bar
```

Would it be valid to provide both config and context of main/to flags?

Reply:
It would indeed!